pages/en/lb4/Creating-components.md

```diff
@@ -307,3 +307,297 @@ class EmailComponent {
+
+export class MyApplication extends Application {
+  super();
+  // Will automatically start when app.start is called
+  this.server(ReportProcessingServer); 
+}
+```
+
+### A more advanced example
+Typically, you'll want server instances that listen for traffic on one or more
+ports (this is why they're called "servers", after all). This leads into a key
+concept to leverage for creating your custom servers.
+
+### Controllers and routing
+LoopBack 4 developers are strongly encouraged to use controllers for their
+modules, and this naturally leads to the concept of routing.
+
+No matter what protocol you intend to use for your custom server, you'll need
+to use some algorithm to determine _which_ controller and function to send
+request data to, and that means you need a router.
+
+For example, consider a "toy protocol" similar to the JSON RPC
+specification (but nowhere near as complete or robust).
+
+The toy protocol will require a JSON payload with three properties: `controller`,
+`method`, and `input`.
+
+An example request would look something like this:
+```json
+{
+  "controller": "GreetController",
+  "method": "basicHello",
+  "input": {
+    "name": "world",
+  }
+}
+```
+
+### Defining the controller
+Here is the `GreetController` and the `Person` type:
+
+{% include code-caption.html content="/src/controllers/greet.controller.ts" %}
+
+```ts
+export class GreetController {
+  basicHello(input: Person) {
+    return `Hello, ${input && input.name || 'World'}!`;
+  }
+
+  hobbyHello(input: Person) {
+    return `${this.basicHello(input)} I heard you like ${input && input.hobby ||
+      'underwater basket weaving'}.`;
+  }
+}
+```
+
+{% include code-caption.html content="/src/models/person.model.ts" %}
+```ts
+// Note that this can also be a class!
+export type Person = {
+  name: string;
+  hobby: string;
+};
+```
+
+### Defining our router and server
+Now that there is a controller to test with, the following code creates a `Router` and `Server`.
+This example uses [express](https://expressjs.com/) and the
+`@types/express` package.
+
+{% include code-caption.html content="/src/server/rpc-router.ts" %}
+```ts
+import {RPCServer} from './rpc-server';
+import * as express from 'express';
+import * as parser from 'body-parser';
+
+export class RPCRouter {
+  routing: express.Router;
+  constructor(private server: RPCServer) {
+    this.routing = express.Router();
+    const jsonParser = parser.json();
+    this.routing.post('*', jsonParser, async (request, response) => {
+      console.log(JSON.stringify(request.body));
+      // Aliasing these values for clarity and brevity.
+      const ctrl = request.body.controller;
+      const method = request.body.method;
+      const input = request.body.input;
+      try {
+        const controller = await this.server.get(`controllers.${ctrl}`);
+        if (!controller) {
+          response.statusCode = 400;
+          response.send(`No controller was found with name "${ctrl}".`);
```

Comment:
I'd personally `throw new HttpErrors.BadRequest`. However, since the controller was not found, it's probably better to send 404? Also, if this RPC is JSON based, shouldn't we return a JSON formatted error, possibly with 200 status code?